includes/class-webmention-avatar-handler.php

```diff
@@ -0,0 +1,170 @@
+		add_filter( 'avatar_defaults', array( $cls, 'anonymous_avatar' ) );
+	}
+
+	public static function anonymous_avatar( $avatar_defaults ) {
+		$avatar_defaults['mystery'] = __( 'Mystery Person (hosted locally)', 'webmention' );
+		return $avatar_defaults;
+	}
+
+	/**
+	 * Function to retrieve Avatar if stored in meta
+	 *
+	 *
+	 * @param int|WP_Comment $comment
+	 *
+	 * @return string $url
+	 */
+	public static function get_avatar_meta( $comment ) {
+		if ( is_numeric( $comment ) ) {
+			$comment = get_comment( $comment );
+		}
+		if ( ! $comment ) {
+			return false;
+		}
+		$avatar = get_comment_meta( $comment->comment_ID, 'avatar', true );
+		// Backward Compatibility for Semantic Linkbacks
+		if ( ! $avatar ) {
+			$avatar = get_comment_meta( $comment->comment_ID, 'semantic_linkbacks_avatar', true );
+		}
+		return $avatar;
+	}
+
+
+	/**
+	 * Function to retrieve default avatar URL
+	 *
+	 *
+	 * @param string $type Default Avatar URL
+	 *
+	 * @return string|boolean $url
+	 */
+	public static function get_default_avatar( $type = null ) {
+		if ( ! $type ) {
+			$type = get_option( 'avatar_default', 'mystery' );
+		}
+		switch ( $type ) {
+			case 'mm':
+			case 'mystery':
+			case 'mysteryman':
+				return plugin_dir_url( dirname( __FILE__ ) ) . 'img/mm.jpg';
+		}
+		return apply_filters( 'webmention_default_avatar', $type );
+	}
+
+	/**
+	 * Function to check if there is a gravatar
+	 *
+	 *
+	 * @param WP_Comment $comment
+	 *
+	 * @return boolean
+	 */
+	public static function check_gravatar( $comment ) {
+		$hash     = md5( strtolower( trim( $comment->comment_author_email ) ) );
+		$url      = 'https://www.gravatar.com/avatar/' . $hash . '?d=404';
+		$response = wp_remote_head( $url );
+		if ( is_wp_error( $response ) || 404 === wp_remote_retrieve_response_code( $response ) ) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Replaces the default avatar with a locally stored default
+	 *
+	 * @param array             $args Arguments passed to get_avatar_data(), after processing.
+	 * @param WP_Comment $comment A comment object
+	 *
+	 * @return array $args
+	 */
+	public static function anonymous_avatar_data( $args, $comment ) {
+		if ( ! $comment instanceof WP_Comment ) {
+			return $args;
+		}
+		$local = apply_filters( 'webmention_local_avatars', array( 'mm', 'mystery', 'mysteryman' ) );
+		if ( ! in_array( $args['default'], $local, true ) ) {
+			return $args;
+		}
+		// Always override if default forced
+		if ( $args['force_default'] ) {
+			$args['url'] = self::get_default_avatar( $args['default'] );
+			return $args;
+		}
+		if ( ! strpos( $args['url'], 'gravatar.com' ) ) {
+			return $args;
+		}
+		$args['url'] = str_replace( 'd=' . $args['default'], 'd=' . self::get_default_avatar( $args['default'] ), $args['url'] );
+		if ( ! empty( $comment->comment_author_email ) ) {
+			if ( self::check_gravatar( $comment ) ) {
+				update_comment_meta( $comment->comment_ID, 'avatar', $args['url'] );
```

Comment:
does this store the local anonymous avatar to the db? I would avoid storing anonymous or temporary data to the db, because that means a lot of migration when we have/want to change the avatar and it makes it impossible to use a different one.

Reply:
Yes, it's how Semantic Linkbacks is doing it now.. but will think of another solution based on that